tests/publish.rs

```diff
@@ -14,111 +14,127 @@
 pub mod test_lib;
 
 use async_std::task;
+use async_trait::async_trait;
 use std::{
     sync::{Arc, Mutex},
     time,
 };
-use up_rust::{Data, UMessage, UMessageBuilder, UPayloadFormat, UStatus, UTransport, UUIDBuilder};
+use test_case::test_case;
+use up_rust::{
+    Data, UListener, UMessage, UMessageBuilder, UPayloadFormat, UStatus, UTransport, UUIDBuilder,
+    UUri,
+};
 
+struct PublishNotificationListener {
+    recv_data: Arc<Mutex<String>>,
+}
+impl PublishNotificationListener {
+    fn new() -> Self {
+        PublishNotificationListener {
+            recv_data: Arc::new(Mutex::new(String::new())),
+        }
+    }
+    fn get_recv_data(&self) -> String {
+        self.recv_data.lock().unwrap().clone()
+    }
+}
+#[async_trait]
+impl UListener for PublishNotificationListener {
+    async fn on_receive(&self, msg: UMessage) {
+        if let Data::Value(v) = msg.payload.unwrap().data.unwrap() {
+            let value = v.into_iter().map(|c| c as char).collect::<String>();
+            *self.recv_data.lock().unwrap() = value;
+        } else {
+            panic!("The message should be Data::Value type.");
+        }
+    }
+    async fn on_error(&self, err: UStatus) {
+        panic!("Internal Error: {err:?}");
+    }
+}
+
+#[test_case(test_lib::create_utransport_uuri(Some(0), 0, 0), test_lib::create_utransport_uuri(Some(0), 0, 0); "Normal UUri")]
+#[test_case(test_lib::create_utransport_uuri(Some(0), 0, 0), test_lib::create_special_uuri(0); "Special UUri")]
 #[async_std::test]
-async fn test_publish_and_subscribe() {
+async fn test_publish_and_subscribe(origin_uuri: UUri, dst_uuri: UUri) {
     test_lib::before_test();
 
     // Initialization
     let target_data = String::from("Hello World!");
-    let upclient = test_lib::create_up_client_zenoh().await.unwrap();
-    let uuri = test_lib::create_utransport_uuri(0);
-    let verified_data = Arc::new(Mutex::new(String::new()));
+    let upclient_send = test_lib::create_up_client_zenoh(0, 0).await.unwrap();
+    let upclient_recv = test_lib::create_up_client_zenoh(1, 1).await.unwrap();
 
     // Register the listener
-    let verified_data_cloned = verified_data.clone();
-    let listener = move |result: Result<UMessage, UStatus>| match result {
-        Ok(msg) => {
-            if let Data::Value(v) = msg.payload.unwrap().data.unwrap() {
-                let value = v.into_iter().map(|c| c as char).collect::<String>();
-                *verified_data_cloned.lock().unwrap() = value;
-            } else {
-                panic!("The message should be Data::Value type.");
-            }
-        }
-        Err(ustatus) => panic!("Internal Error: {ustatus:?}"),
-    };
-    let listener_string = upclient
-        .register_listener(uuri.clone(), Box::new(listener))
+    let pub_listener = Arc::new(PublishNotificationListener::new());
+    upclient_recv
+        .register_listener(dst_uuri.clone(), pub_listener.clone())
```

Comment:
```suggestion
        .register_listener(origin_uuri.clone(), pub_listener.clone())
```
?

Reply:
Yes, this one. Thank you for pointing out. I think I should rename `dst_uuri` to `listen_uuri` for better understanding.